### DISCUSSION TITLE
Proper usage of RpcClient.write()

### DISCUSSION THREAD

**Discussion by paliasgh** [robotology/community#710] [Q&A] [answered]
Hello, 

I hope you are going great. We have a YARP related question again. We need to periodically check if robot is still speaking or not. Therefore we connect to the rpc port `/iSpeak/rpc` in our code with:

`RpcClient speakCheckPort;`

And do this in `updateModule()`:

```
Bottle cmd, reply;
cmd.addVocab32(Vocab32::encode("stat"));
speakCheckPort.write(cmd,reply);
string msg=reply.get(0).asString().c_str();
if(msg == "quiet") {talking=false;}
```
The problem is, when the robot is saying a long sentence, `/iSpeak/rpc` will not respond right away. It waits until the robot stops talking and then responds  "quiet". The line `speakCheckPort.write(cmd,reply);` will block the entire `updateModule()` waiting for a reply from `/iSpeak/rpc`. Is there any way the code can send a request to `/iSpeak/rpc` and if there is no response right away, pass and keep doing it until there is a response?

Thank you!

**Comment by pattacini** [comment]
Hi @paliasgh 

I'm a bit rusty with YARP and iSpeak but just a quick try for you.
Without changing the server part, you could set a timeout:

```c++
yarp::os::RpcClient rpcPort;
// ... (prepare your RPC command)

rpcPort.setTimeout(1.0);  // Set 1-second timeout
auto ret = rpcPort.write(cmd, reply);

if (reply.isValid() && ret) {  
    // Process the 'reply'
} else {
    // Handle timeout or error
}
```

**Reply by paliasgh** [reply]
Hi @pattacini ,

Thank you for your prompt response. It seems like there is no `setTimeout()` in `RpcClient` class, even looking at [what is available 
](https://yarp.it/latest//classyarp_1_1os_1_1RpcServer.html). Maybe the only solution is to modify the iSpeak server's behaviour (which can be hard)?

```
error: ‘class yarp::os::RpcClient’ has no member named ‘setTimeout’
         speakCheckPort.setTimeout(0.1);
```
--
By the way, in the [documentation](https://yarp.it/latest//classyarp_1_1os_1_1RpcServer.html#a787b7dbb52d7cb55c42d6a716c08baea), should this `write()` function and also some other functions be under `RpcClient`, not the `RpcServer`?

**Reply by pattacini** [reply, marked as answer]
>Thank you for your prompt response. It seems like there is no setTimeout() in RpcClient class

My bad. You should be able to get access to it this way:

```c++
rpcPort.asPort().setTimeout(1.0);
```

**Reply by paliasgh** [reply]
Hello @pattacini , thank you very much! It solved the problem.
